lib/auth/userloginstate/generator_test.go

```diff
@@ -0,0 +1,254 @@
+				}),
+			},
+			roles: []string{"role1", "role2"},
+			expected: newUserLoginState(t, "user",
+				[]string{
+					"role1",
+				}, trait.Traits{
+					"trait1": []string{"value1"},
+				}),
+		},
+		{
+			name: "access lists add roles with duplicates",
+			accessLists: []*accesslist.AccessList{
+				newAccessList(t, "1", []string{"user"}, []string{"role1", "role2"}, trait.Traits{}),
+				newAccessList(t, "2", []string{"user"}, []string{"role2", "role3"}, trait.Traits{}),
+			},
+			roles: []string{"role1", "role2", "role3"},
+			expected: newUserLoginState(t, "user",
+				[]string{
+					"role1",
+					"role2",
+					"role3",
+				}, trait.Traits{}),
+		},
+		{
+			name: "access lists add traits with duplicates",
+			accessLists: []*accesslist.AccessList{
+				newAccessList(t, "1", []string{"user"}, []string{},
+					trait.Traits{
+						"trait1": []string{"value1", "value2"},
+						"trait2": []string{"value3", "value4"},
+					},
+				),
+				newAccessList(t, "2", []string{"user"}, []string{},
+					trait.Traits{
+						"trait2": []string{"value3", "value1"},
+						"trait3": []string{"value5", "value6"},
+					},
+				),
+			},
+			expected: newUserLoginState(t, "user", []string{},
+				trait.Traits{
+					"trait1": []string{"value1", "value2"},
+					"trait2": []string{"value3", "value4", "value1"},
+					"trait3": []string{"value5", "value6"},
+				}),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := context.Background()
+			svc, backendSvc := initSvc(t)
+
+			for _, accessList := range test.accessLists {
+				_, err = backendSvc.UpsertAccessList(ctx, accessList)
+				require.NoError(t, err)
+			}
+
+			for _, role := range test.roles {
+				role, err := types.NewRole(role, types.RoleSpecV6{})
+				require.NoError(t, err)
+				require.NoError(t, backendSvc.UpsertRole(ctx, role))
+			}
+
+			state, err := svc.Generate(ctx, user)
+			require.NoError(t, err)
+			require.Empty(t, cmp.Diff(test.expected, state,
+				cmpopts.SortSlices(func(str1, str2 string) bool {
+					return str1 < str2
+				})))
+		})
+	}
+}
+
+type svc struct {
+	services.AccessLists
+	services.Access
+}
+
+func initSvc(t *testing.T) (*Generator, *svc) {
+	t.Helper()
+
+	clock := clockwork.NewFakeClock()
+	mem, err := memory.New(memory.Config{
+		Clock: clock,
+	})
+	require.NoError(t, err)
+
+	accessListsSvc, err := local.NewAccessListService(mem, clock)
+	require.NoError(t, err)
+	accessSvc := local.NewAccessService(mem)
+
+	return &Generator{accessLists: accessListsSvc, roles: accessSvc, clock: clock},
+		&svc{AccessLists: accessListsSvc, Access: accessSvc}
+}
+
+func newAccessList(t *testing.T, name string, members []string, roles []string, traits trait.Traits) *accesslist.AccessList {
+	t.Helper()
+
+	alMembers := make([]accesslist.Member, len(members))
+	for i, member := range members {
+		alMembers[i] = accesslist.Member{
+			Name:    member,
+			Joined:  time.Now(),
+			Expires: time.Now().Add(24 * time.Hour),
```

Comment:
I wonder if we should consider the expiry time when issuing certificates. Otherwise, you can issue a certificate 5 minutes before someone's membership expires, but they can keep using it for as long as the certificate is valid: potentially a long time.

Reply:
IIRC we've actually got this issue as with access requests and other things -- I'm unsure of the right thing to do here either. I'd love to get more opinions, perhaps @espadolini?

Reply:
For access requests we've only recently added an explicit flag to let them be issued for longer than the original credentials; in general, the time range in which anything derived from an identity is valid should be contained in the original identity's validity unless otherwise explicitly specified, IMO.

Reply:
Would you be amenable to addressing this in a follow on? I'm a little worried about making this PR more complicated than it already is.

Reply:
Fair enough, I think such changes would naturally apply on top of the current patchset.

cc @jentfoo for awareness, this issue may benefit from deeper thought or a follow-up audit.